internal/scheduler/legacyscheduler.go

```diff
@@ -318,151 +400,180 @@ func uuidFromUlidString(ulid string) (uuid.UUID, error) {
+	// Each gang is made up of a slice of jobs with equal value for gangIdAnnotation.
+	gangsByQueue map[string]map[string][]*JobSchedulingReport
+}
+
+func NewQueueCandidateGangIterator(ctx context.Context, it *CandidateJobsIterator, gangIdAnnotation, gangCardinalityAnnotation string) *CandidateGangIterator {
+	return &CandidateGangIterator{
+		ctx:                       ctx,
+		CandidateJobsIterator:     *it,
+		gangIdAnnotation:          gangIdAnnotation,
+		gangCardinalityAnnotation: gangCardinalityAnnotation,
+		gangsByQueue:              make(map[string]map[string][]*JobSchedulingReport),
 	}
+}
 
-	// Try to find a node for this pod.
-	// Store the report returned by the NodeDb.
-	req := schedulerobjects.PodRequirementsFromPodSpec(podSpec, sched.SchedulingConfig.Preemption.PriorityClasses)
-	var report *PodSchedulingReport
-	var err error
-	if bind {
-		report, err = sched.NodeDb.SelectAndBindNodeToPod(jobId, req)
-	} else {
-		report, err = sched.NodeDb.SelectNodeForPod(jobId, req)
+func (it *CandidateGangIterator) Next() ([]*JobSchedulingReport, error) {
+	// Once this function has returned error,
+	// it will return this error on every invocation.
+	if it.err != nil {
+		return nil, it.err
 	}
-	if err != nil {
-		return nil, err
+
+	// Get one job at a time from the underlying iterator until we either
+	// 1. get a job that isn't part of a gang, in which case we yield it immediately, or
+	// 2. get the final job in a gang, in which case we yield the entire gang.
+	for {
+		report, err := it.CandidateJobsIterator.Next()
+		if err != nil {
+			return nil, err
+		}
+		if report == nil {
+			return nil, nil
+		}
+
+		gangId, gangCardinality, isGangJob, err := GangIdAndCardinalityFromJob(report.Job, it.gangIdAnnotation, it.gangCardinalityAnnotation)
+		if err != nil {
+			log := ctxlogrus.Extract(it.ctx)
+			logging.WithStacktrace(log, err).Errorf("failed to get gang cardinality for job %s", report.Job.Id)
+			gangCardinality = 1 // Schedule jobs with invalid gang cardinality one by one.
+		}
+		if isGangJob {
+			jobsByGangId, ok := it.gangsByQueue[report.Job.Queue]
+			if !ok {
+				jobsByGangId = make(map[string][]*JobSchedulingReport)
+				it.gangsByQueue[report.Job.Queue] = jobsByGangId
+			}
+
+			jobsByGangId[gangId] = append(jobsByGangId[gangId], report)
+			gang := jobsByGangId[gangId]
+			if len(gang) == gangCardinality {
+				delete(jobsByGangId, gangId)
+				return gang, nil
+			}
+		} else {
+			return []*JobSchedulingReport{report}, nil
+		}
 	}
-	return report, nil
+}
+
+func (sched *LegacyScheduler) podRequirementsFromJobs(jobs []*api.Job) []*schedulerobjects.PodRequirements {
+	rv := make([]*schedulerobjects.PodRequirements, 0, len(jobs))
+	for _, job := range jobs {
+		rv = append(rv, sched.podRequirementsFromJob(job)...)
+	}
+	return rv
+}
+
+func (sched *LegacyScheduler) podRequirementsFromJob(job *api.Job) []*schedulerobjects.PodRequirements {
+	rv := make([]*schedulerobjects.PodRequirements, 0, 1+len(job.PodSpecs))
```

Comment:
this is a weird function:
- It allows multiple pod specs (which we don't elsewhere)
- It allows both podspec and podspecs (which we don't elsewhere)

Reply:
I agree that it's a bit weird. We may want to leave it as-is for now since it'd be good to be moving towards a scheduler that supports multi-pod jobs.